### ISSUE TITLE
Inconsistent Model Loading with `weights_only=True` for Custom Models

### ISSUE THREAD

**Issue by andrewrgarcia** [pytorch/pytorch#140230]
### 🚀 The feature, motivation and pitch

I'm encountering an issue with the `weights_only=True` parameter in `torch.load`, where I need to load custom model classes while ensuring repeatable results across multiple runs. My current workaround involves setting `weights_only=False` to achieve identical predictions, which raises security warnings due to potential arbitrary code execution. This issue serves to inform the PyTorch team about the challenges in achieving both secure and repeatable model loading in use cases like neural architecture search, where multiple models are saved and loaded regularly.

**Background**:

In my work with neural architecture search, achieving repeatability is essential. When loading a saved custom model using `weights_only=True`, I encounter issues with loading state and repeatability, even after adding custom classes to `torch.serialization.add_safe_globals`. Using `weights_only=False` resolves the issue, but it raises security concerns. Given the importance of security and repeatability, this feature request aims to address these limitations and propose an improvement for handling custom models in `weights_only=True` mode.

### Alternatives

The current workaround involves setting `weights_only=False`, which allows repeatability but may pose a security risk due to the potential execution of arbitrary code during unpickling. However, this approach is not ideal for secure applications, especially when working with multiple models in neural architecture search.

### Additional context

Environment details:
- PyTorch version: 2.1.0
- Python version: 3.10
- OS: Ubuntu 22.04

Example code and detailed reproduction steps are provided below to illustrate the issue and expected vs. actual behavior.

Thank you for your time and assistance. Any guidance or recommendations would be greatly appreciated.

```
import torch
import numpy as np
import random
from my_custom_package import NeuralNetwork, Trainer  # Import custom model and trainer

# Set random seed for reproducibility
def set_random_seed(seed=42):
    torch.manual_seed(seed)
    np.random.seed(seed)
    random.seed(seed)
    if torch.cuda.is_available():
        torch.cuda.manual_seed(seed)
        torch.cuda.manual_seed_all(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

set_random_seed(42)

# Initialize and train the model
model = NeuralNetwork(input_size=10, output_size=4, num_vars=40)
trainer = Trainer(model=model, lr=1e-4, batch_size=20, num_epochs=20)
trainer.fit(X_train, y_train, X_val, y_val)

# Save the trained model
trainer.save_model('neural_model.pth')

# Function to generate predictions
def generate_predictions(model, X_test):
    with torch.no_grad():
        X_test_tensor = torch.tensor(X_test, dtype=torch.float32)
        return model(X_test_tensor)

# Generate predictions with the original model
predictions = generate_predictions(model, X_test)

# Load the model with weights_only=False and evaluate
loaded_model = torch.load('neural_model.pth', weights_only=False)
loaded_model.eval()
predictions_loaded_model = generate_predictions(loaded_model, X_test)

# Load the model again with default settings and evaluate
loaded_model_2 = torch.load('neural_model.pth')
loaded_model_2.eval()
predictions_loaded_model_2 = generate_predictions(loaded_model_2, X_test)

# Check consistency of predictions
print("All Same - trained and loaded:\t", np.allclose(predictions, predictions_loaded_model))
print("All Same - loaded once, loaded twice:\t", np.allclose(predictions_loaded_model_2, predictions_loaded_model_2))

```

**Comment by malfet** [comment]
@andrewrgarcia do you mind sharing `neural_model.pth'? Or modify your test to compare loaded weights? 
Another question: if you load the same model twice with `weights_only=False` would it result in the same predictions? (I.e. could it be that your model involves something like dropout and RNG state between two invocations changes)

**Comment by mikaylagawarecki** [comment]
@malfet Added a small test case here https://github.com/pytorch/pytorch/pull/140850, not sure if this was what you were looking for 

fwiw, Dropout is a module that does not carry state, so I don't think Dropout ratio might be what is at play here

```python
>>> m = torch.nn.Dropout(0.1)
>>> m.state_dict()
OrderedDict()
```

**Comment by mikaylagawarecki** [comment]
@andrewrgarcia would you be willing to share your `neural_model.pth`?  that would be very helpful for debugging

**Comment by andrewrgarcia** [comment]
> @andrewrgarcia would you be willing to share your `neural_model.pth`?  that would be very helpful for debugging

@malfet Hello Yes. Sorry for the delay. I am away from my computer and will return this Monday, Tuesday at the very latest to give you an update / the information requested. Many thanks!

**Comment by mikaylagawarecki** [comment]
Hey @andrewrgarcia, gentle bump on this one if you have the `neural_model.pth` file available (sorry to rush you, quite eager to debug this one, thanks so much :))

**Comment by andrewrgarcia** [comment]
Dear PyTorch Team,  

Thank you so much for your attention to the issue I previously opened. After revisiting my workflow and investigating further, I believe the confusion may have stemmed from my own uncertainty about the correct protocol to use when saving and loading models—particularly when working with multiple automatically generated  architectures (as is the case when doing NAS).  

I experimented with both approaches (`entire_model=True` and `entire_model=False`) and wrote the following code to test the consistency of predictions between trained and loaded models. My current understanding is that using `entire_model=False` (saving and loading weights only) is the safer and more robust approach when dealing with diverse architectures, as it decouples the architecture definition from the state.  

Here is the code I used for testing:  

```python
def model_loading_tests(entire_model=True):
    '''
    Test loading a saved PyTorch model by comparing predictions.

    Parameter
    ------------
    entire_model (bool): 
        If True, loads the entire model (architecture + weights) using `torch.load()`, which is convenient but may pose a security risk if the file is untrusted. If False, loads only the model's weights using `load_state_dict()`.
    '''
    model = start_model_train_n_save('neural_model.pth', entire_model)
    predictions = generate_predictions(model)

    # Load the model based on the chosen method and ensure it's in eval mode
    if entire_model:
        # Load the entire model
        loaded_model = torch.load('neural_model.pth', weights_only=False)
        loaded_model.eval()  # Ensure evaluation mode for inference
    else:
        # Load only weights into the same model architecture
        model.load_state_dict(torch.load('neural_model.pth', weights_only=True))
        loaded_model = model

    predictions_loaded_model = generate_predictions(loaded_model)

    # Test reloading the model again to verify consistency
    if entire_model:
        loaded_model_2 = torch.load('neural_model.pth', weights_only=False)
        loaded_model_2.eval()  # Ensure evaluation mode for inference
    else:
        model.load_state_dict(torch.load('neural_model.pth', weights_only=True))
        loaded_model_2 = model

    predictions_loaded_model_2 = generate_predictions(loaded_model_2)

    print("Predictions Are the Same - trained & loaded models:\t\t", np.allclose(predictions, predictions_loaded_model))
    print("Predictions Are the Same - model loaded once & loaded twice:\t", np.allclose(predictions_loaded_model, predictions_loaded_model_2))

model_loading_tests(entire_model=False)
model_loading_tests(entire_model=True)
```  

Additionally, as requested, I am attaching the `neural_model.pth` file for reference (with .txt extension because .pth is not accepted by gh) as well as its computational graph in graphical form and txt.  

Could you kindly confirm if this approach (particularly `entire_model=False`) aligns with best practices for saving and loading models in PyTorch, especially when working with dynamically generated architectures?  

Thank you again for your time and attention, and for all the incredible work you do to support the community. I truly appreciate your help and guidance.  

Best regards,  
Andrew Garcia

[neural_model.pth.txt](https://github.com/user-attachments/files/17805985/neural_model.pth.txt)
[torch-graph-metaconv.pdf](https://github.com/user-attachments/files/17806000/torch-graph-metaconv.pdf)
[torch-graph-metaconv.txt](https://github.com/user-attachments/files/17806026/torch-graph-metaconv.txt)

**Comment by mikaylagawarecki** [comment]
Hi @andrewrgarcia, thanks for sharing your checkpoint and script :)

To confirm

> I believe the confusion may have stemmed from my own uncertainty about the correct protocol to use when saving and loading models—particularly when working with multiple automatically generated architectures (as is the case when doing NAS).

Do you mean that after you revisited this there was no longer a discrepancy between `weights_only=True` and `weights_only=False`? And there was no longer an issue with repeatibility for `weights_only=True`?

I notice that the `neural_model.pth` you shared contained just the state_dict and this script ran successfully with it, so it seems there is actually no discrepancy between `weights_only=True` and `weights_only=False`, is this accurate?

```python
import torch

wo_true = torch.load("neural_model.pth")
wo_false = torch.load("neural_model.pth", weights_only=False)
for k in wo_true.keys():
    assert torch.equal(wo_true[k], wo_false[k])
```

> Could you kindly confirm if this approach (particularly entire_model=False) aligns with best practices for saving and loading models in PyTorch,

Indeed saving the `state_dict` of a model is a recommended best practice. See [the documentation here](https://pytorch.org/docs/main/notes/serialization.html#saving-and-loading-torch-nn-modules) where we recommend the following

> Instead of saving a module directly, for compatibility reasons it is recommended to instead save only its state dict. Python modules even have a function, [load_state_dict()](https://pytorch.org/docs/main/generated/torch.nn.Module.html#torch.nn.Module.load_state_dict), to restore their states from a state dict:

Further, loading a `state_dict` is the intended use case for `weights_only=True`.
